src/machine/machine_atmega328p.go

```diff
@@ -467,3 +472,93 @@ var SPI0 = SPI{
 	sdo:  PB3,
 	sdi:  PB4,
 	cs:   PB2}
+
+// Pin Change Interrupts
+var (
+	pinCallbacks [3][8]func(Pin)
+	interrupts   [3]*interrupt.Interrupt
+	previous     [3]uint8
+)
+
+func handlePCINTInterrupts(intr uint8, port *volatile.Register8) {
+	current := port.Get()
+	for i := uint8(0); i < 8; i++ {
+		if (current>>i)&0x01 == 0x01 && pinCallbacks[intr][i] != nil {
+			pinCallbacks[intr][i](Pin(i * (intr + 1)))
+		}
+	}
+	previous[intr] = current
+}
+
+func handlePCINT0Interrupts(intr interrupt.Interrupt) {
+	handlePCINTInterrupts(avr.IRQ_PCINT0-avr.IRQ_PCINT0, avr.PINB)
+}
+
+func handlePCINT1Interrupts(intr interrupt.Interrupt) {
+	handlePCINTInterrupts(avr.IRQ_PCINT1-avr.IRQ_PCINT0, avr.PINC)
+}
+
+func handlePCINT2Interrupts(intr interrupt.Interrupt) {
+	handlePCINTInterrupts(avr.IRQ_PCINT2-avr.IRQ_PCINT0, avr.PIND)
+}
+
+func SetInterrupt(pin Pin, callback func(Pin)) (err error) {
+	if callback == nil {
+		// Disable this pin interrupt
+		return ErrFeatureNotImplemented
+	}
+	var PCMSK_REG *volatile.Register8
+	var PCIE uint8
+	var PCINT uint8
+	var IRQ uint8
+	var PIN0 Pin
+
+	switch {
+	case pin >= PB0 && pin <= PB7:
+		// PCMSK0 - PCINT0-7
+		PCMSK_REG = avr.PCMSK0
+		PCIE = avr.PCICR_PCIE0
+		PIN0 = PD0
+		PCINT = 1 << (pin - PIN0)
+		IRQ = avr.IRQ_PCINT0
+		if interrupts[IRQ-avr.IRQ_PCINT0] == nil {
+			intr := interrupt.New(avr.IRQ_PCINT0, handlePCINT0Interrupts)
+			interrupts[IRQ-avr.IRQ_PCINT0] = &intr
+		}
+	case pin >= PC0 && pin <= PC7:
+		// PCMSK1 - PCINT8-14
+		PCMSK_REG = avr.PCMSK1
+		PCIE = avr.PCICR_PCIE1
+		PIN0 = PC0
+		PCINT = 1 << (pin - PIN0)
+		IRQ = avr.IRQ_PCINT1
+		if interrupts[IRQ-avr.IRQ_PCINT0] == nil {
+			intr := interrupt.New(avr.IRQ_PCINT1, handlePCINT1Interrupts)
+			interrupts[IRQ-avr.IRQ_PCINT0] = &intr
+		}
+	case pin >= PD0 && pin <= PD7:
+		// PCMSK2 - PCINT16-23
+		PCMSK_REG = avr.PCMSK2
+		PCIE = avr.PCICR_PCIE2
+		PIN0 = PD0
+		PCINT = 1 << (pin - PIN0)
+		IRQ = avr.IRQ_PCINT2
+		if interrupts[IRQ-avr.IRQ_PCINT0] == nil {
+			intr := interrupt.New(avr.IRQ_PCINT2, handlePCINT2Interrupts)
+			interrupts[IRQ-avr.IRQ_PCINT0] = &intr
+		}
+	default:
+		return ErrInvalidInputPin
+	}
+
+	if pinCallbacks[IRQ-avr.IRQ_PCINT0][pin-PIN0] != nil {
+		return ErrCallbackAlreadyDefined
```

Comment:
It depends on the chip, but at least on the nrf a new call to `SetInterrupt` simply overwrites the old interrupt. (There is a risk of a race condition in the write though, so perhaps better to leave it as-is).

Reply:
I don't remember where I saw this idea, but I was translating from something... either one of the other tinygo architectures or something in the arduino framework.
I can see the possibility of wanting to change the interrupt handler at runtime, but I don't personally have a use case, so I thought I'd leave it like this until there's a good reason to implement.